### PR TITLE
[zero_to_fp32.py] support param groups

### PR DESCRIPTION
**Pull request by stas00**
In my original version I happened to use a model with 1 param group so I wasn't aware that there could be multiple flattened tensors - one per group, so my reconstruction script [was breaking]( https://github.com/microsoft/DeepSpeed/issues/1009) when it run into non-one-single-flat-tensor.

This PR tries to fix that.

There might be a more efficient way to do it, but for now just trying to make sure that it functions correctly.

I also left some disabled debug code for now while it's new and likely to need debug still. We can remove it later if we feel it's solid.

While I tested this on a few live models, it'd be great to have a functional test for zero2 and zero3 for this code. But I'm not quite familiar with how your test suite is done and unfortunately don't really have time right now to sort it out.

The simplest conceptual test would be:
```
# original_fp32_state_dict_path
model = (multi-param-group model).from_pretrained(original_fp32_state_dict_path)
engine = deepspeed.initialize(model, ...)
engine.save_checkpoint()
! ./zero_to_fp32.py global_step1 pytorch_model.bin
! diff pytorch_model.bin original_fp32_state_dict_path # should be identical
```

Fixes: https://github.com/microsoft/DeepSpeed/issues/1009

@exelents, please check that this PR solves the problem for you.

### CODE REVIEW

File: deepspeed/utils/zero_to_fp32.py [hunk 1]
```diff
@@ -50,12 +54,15 @@ def parse_optim_states(files):
...
+    # if there is more than one param group, there will be multiple flattened tensors - one
+    # flattened tensor per group - for simplicity merge them into a single tensor
+    #
+    # XXX: could make the script more memory efficient for when there are multiple groups - it
+    # will require matching the sub-lists of param_shapes for each param group flattened tensor
     fp32_flat_groups = [
-        state_dicts[i]['optimizer_state_dict'][fp32_groups_key][0]
-        for i in range(len(state_dicts))
+        torch.cat(state_dicts[i]['optimizer_state_dict'][fp32_groups_key],
+                  0) for i in range(len(state_dicts))
```
Comment (stas00) [inline deepspeed/utils/zero_to_fp32.py:64]: This is the only functional change in this PR. Instead using just the first element, it now uses them all.

**Comment by exelents** [comment]
Converting of "siamese" model based on t5-11b encoders is done successfully. But when I load it into CPU memory I gone a stange message:
```
Some weights of T5Siamese were not initialized from the model checkpoint at ./siamese_train_deepspeed/models/siamese-t5-11b-fp16/checkpoint-1625 and are newly initialized: ['encoder_left.encoder.embed_tokens.weight', 'encoder_right.encoder.embed_tokens.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
```

File: deepspeed/utils/zero_to_fp32.py (hunk 1 above)
Comment (tjruwase) [reply deepspeed/utils/zero_to_fp32.py:64]: This seems fine for now. I agree we have to revisit, especially for very large models that could cause CPU OOM.

**Review by tjruwase — APPROVED** [review]

**Comment by stas00** [comment]
Thank you for running the checks, @exelents 

> ./siamese_train_deepspeed/models/siamese-t5-11b-fp16/checkpoint-1625 and are newly initialized: 
> ['encoder_left.encoder.embed_tokens.weight', 'encoder_right.encoder.embed_tokens.weight']

You have `encoder_right.shared.weight` - aren't those tied/aliased? (Same for left)

**Comment by exelents** [comment]
> You have `encoder_right.shared.weight` - aren't those tied/aliased? (Same for left)

Yes, I have these weights returned from named_parameters() function of my model, so I suppose they should exist in the checkpoint.

```
dd = list(model_engine.module.named_parameters())
dd = list(filter(lambda x: 'shared' in x[0], dd))
dd
```
```
[('encoder_left.shared.weight',
  Parameter containing:
  tensor([1.], device='cuda:0', dtype=torch.float16, requires_grad=True)),
 ('encoder_right.shared.weight',
  Parameter containing:
  tensor([1.], device='cuda:0', dtype=torch.float16, requires_grad=True))]
```

**Comment by stas00** [comment]
But that's what I'm saying: the checkpoint does have `encoder_left.shared.weight` and `encoder_right.shared.weight' and those are restored. 

The loader complains about `encoder_left.encoder.embed_tokens.weight` and `encoder_right.encoder.embed_tokens.weight`.

In your `dd` dump above they don't exist.

BTW, the new version of `zero_to_fp32.py` has a `debug` flag - turn it on and when you run it you will see each weight as it gets loaded.

**Comment by exelents** [comment]
I turned on debug flag and it show me all weights in checkpoint including:
└──>$ cat debug.txt | grep shared
encoder_left.shared.weight full shape: torch.Size([32128, 512]) partition0 numel=16449536 partitioned_padding_numel=0
encoder_right.shared.weight full shape: torch.Size([32128, 512]) partition0 numel=16449536 partitioned_padding_numel=0

**Comment by stas00** [comment]
We have already established that.

Please review https://github.com/microsoft/DeepSpeed/pull/1017#issuecomment-829120612 the warning is for 2 other names.

**Comment by exelents** [comment]
Ah, okay. I understood. In the code of T5EncoderModel I see that variable encoder_right.encoder.embed_tokens is initialized from external variable encoder_left.shared while T5Stack is created, and it isn't needed to be saved.
Thank you.

**Comment by stas00** [comment]
Yes, probably could clean that up so that it doesn't produce a misleading warning.

The key is to please check that the resumed checkpoint scores well for you. I did only a quick 100 or so steps and the loss looked correct. Also please re-check with zero2. I did test it as well, but a second pair of eyes is always better.

I was just concerned that perhaps somehow the saved weights weren't in the same order as the `param_names` dict, so it'd always reshape it correctly, since it's the same number of elements once it's all flattened into a single tensor, but I checked that the order appears to be correct.
